lib/ferrum/browser/tracing.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+module Ferrum
+  class Browser
+    # chrome://tracing
+    class Tracing
+      DEFAULT_CATEGORIES = %w[
+        -*
+        devtools.timeline
+        v8.execute
+        disabled-by-default-devtools.timeline
+        disabled-by-default-devtools.timeline.frame
+        toplevel
+        blink.console
+        blink.user_timing
+        latencyInfo
+        disabled-by-default-devtools.timeline.stack
+        disabled-by-default-v8.cpu_profiler
+        disabled-by-default-v8.cpu_profiler.hires
+      ]
+
+      def initialize(client:)
+        self.client = client
+      end
+
+      def start(path: '', screenshots: false, categories: DEFAULT_CATEGORIES)
+        raise "Cannot start recording trace while already recording trace." if recording
+        categories = categories.concat(["disabled-by-default-devtools.screenshot"]) if screenshots
+        self.path = path
+        self.recording = true
+        @client.command("Tracing.start", transferMode: "ReturnAsStream", categories: categories.join(','))
+      end
+
+      def stop
+        @client.on("Tracing.tracingComplete", handler_type: :once) do |event|
```

Comment:
@route this event is not triggered for some of the reasons, could you put the advice, please?

Reply:
Because test finishes faster than the event comes, and you most likely need @client for page not for the whole browser.

Reply:
Oh wow :tada:, indeed, I didn’t think about it, I thought it’s a matter of configuration. @route Thank you! :muscle:

> page not for the whole browser, because each page should be traced independently

yeah, it makes sense, let's use `page`.

Just a note about `page`:
I see in playwright implementation for `page` reference, for example:
```js
this._tracingClient = page ? (page._delegate as CRPage)._mainFrameSession._client : this._session;
```
Source: https://github.com/microsoft/playwright/blob/c373986ca0b91b76d4910ce0617b55e50934114c/packages/playwright-core/src/server/chromium/crBrowser.ts#L240

however for `puppeteer`, they use the [connection](https://github.com/puppeteer/puppeteer/blob/040a0e561b4f623f7929130b90be129f94ebb642/src/common/Connection.ts) with no specific `page` reference:
```js
export class Tracing {
  _client: CDPSession;
```
Source: https://github.com/puppeteer/puppeteer/blob/040a0e561b4f623f7929130b90be129f94ebb642/src/common/Tracing.ts#L45

Reply:
About:
> each page should be traced independently

```js
  it('should throw if tracing on two pages', async () => {
    await page.tracing.start({ path: outputFile });
    const newPage = await browser.newPage();
    let error = null;
    await newPage.tracing
      .start({ path: outputFile })
      .catch((error_) => (error = error_));
    await newPage.close();
    expect(error).toBeTruthy();
    await page.tracing.stop();
  });
```
Source: https://github.com/puppeteer/puppeteer/blob/040a0e561b4f623f7929130b90be129f94ebb642/test/tracing.spec.ts#L93-L103

So, we probably should avoid using tracing for a few pages at the same time. @route What do you think?

Reply:
Maybe, I don't know because I need to start playing around to prove it. I trust you